src/bot/mtproto-api/workrooms/close-chat.ts

```diff
@@ -0,0 +1,139 @@
+import bigInt from "big-integer";
+
+export async function closeChat(context: Context<"issues.closed", SupportedEvents["issues.closed"]>): Promise<CallbackResult> {
+  const {
+    payload,
+    adapters: {
+      supabase: { chats },
+    },
+    logger,
+  } = context;
+  const mtProto = new MtProto(context);
+  await mtProto.initialize();
+
+  logger.info("Closing chat with name: ", { chatName: payload.issue.title });
+  const chat = await chats.getChatByTaskNodeId(payload.issue.node_id);
+
+  const fetchChat = await mtProto.client.invoke(
+    new mtProto.api.messages.GetFullChat({
+      chatId: chat.chatId,
+    })
+  );
+
+  if (!fetchChat) {
+    throw new Error("Failed to fetch chat");
+  }
+
+  let chatParticipants;
+
+  if ("participants" in fetchChat.fullChat) {
+    chatParticipants = fetchChat.fullChat.participants;
+  } else {
+    throw new Error("Failed to fetch chat participants");
+  }
+
+  // archive it
+  await mtProto.client.invoke(
+    new mtProto.api.folders.EditPeerFolders({
+      folderPeers: [
+        new mtProto.api.InputFolderPeer({
+          peer: new mtProto.api.InputPeerChat({ chatId: chat.chatId }),
+          folderId: 1, // 0 is active, 1 is archived
+        }),
+      ],
+    })
+  );
+
+  if (chatParticipants.className === "ChatParticipants") {
+    await mtProto.client.invoke(
+      new mtProto.api.messages.SendMessage({
+        message: "This task has been closed and this chat has been archived.",
+        peer: new mtProto.api.InputPeerChat({ chatId: chat.chatId }),
+      })
+    );
+
+    const participants = chatParticipants.participants;
+    let creatorId;
+
+    const userIds = participants
+      .map((participant) => {
+        if (participant.className === "ChatParticipantCreator") {
+          creatorId = participant.userId;
+          return undefined;
+        }
+        return participant.userId;
+      })
+      .filter((id) => id !== undefined);
+
+    if (!creatorId) {
+      throw new Error("Failed to get chat creator");
+    }
+
+    userIds.push(creatorId);
+    const chatInput = await mtProto.client.getInputEntity(chat.chatId);
+
+    await chats.userSnapshot(
+      chat.chatId,
+      userIds.map((id) => id.toJSNumber())
+    );
+
+    let generator = deleteChatUsers(mtProto, userIds, context, chatInput);
+    let result = await generator.next();
+
+    while (!result.done) {
+      if (!result.value.success) {
+        const seconds = result.value.error?.seconds ?? 300;
+        logger.info("Retrying...", { seconds, er: result.value.error });
+
+        await intervalLogger(seconds, 60, logger, new Promise((resolve) => setTimeout(resolve, seconds * 1000)));
+
+        generator = deleteChatUsers(mtProto, userIds.slice(result.value.index), context, chatInput);
+      }
+      result = await generator.next();
+    }
+  }
+
+  await chats.updateChatStatus("closed", payload.issue.node_id);
+  return { status: 200, reason: "chat_closed" };
+}
+
+async function* deleteChatUsers(
```

Comment:
Why generator?

Reply:
I'm trying to up my game hahaha, I've written them before then removed in favour of a more basic method but it felt effective to use one here and works a treat. 

Essentially I didn't want the userId snapshot to run again in the event that we did get rate limited or it would have less than the first snapshot and had to handle rate limiting.